lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -65,6 +88,232 @@ static opt<std::string> ABIAnalysisOutputPath("abi-analysis-output",
+    OFPClobbering.addFnAttribute(Attribute::ReadOnly);
+    OFPClobbering.addFnAttribute(Attribute::NoUnwind);
+
+    auto *OpqFunc = OFPClobbering.get(BB->getParent()->getName(),
+                                      Type::getInt64Ty(Context),
+                                      {},
+                                      "opq_func_for_clobbering");
+    for (GlobalVariable *Reg : ClobberedRegs)
+      new StoreInst(CallInst::Create(OpqFunc, "", Br), Reg, Br);
+
+    break;
+  }
+
+  case FunctionKind::NoReturn: {
+    ReturnInst *Ret = ReturnInst::Create(Context);
+    ReplaceInstWithInst(Term, Ret);
+    break;
+  }
+
+  case FunctionKind::Fake: {
+    BranchInst *Br = BranchInst::Create(GCBI::getFallthrough(Term));
+
+    // Get the fake function by its entry basic block.
+    Function *FakeFunction = Oracle.getFakeFunction(Next);
+
+    // Must have already been analyzed.
+    revng_assert(FakeFunction != nullptr);
+
+    // If possible, inline the fake function.
+    auto *CI = CallInst::Create(FakeFunction, "", Term);
+    InlineFunctionInfo IFI;
+    bool Status = InlineFunction(llvm::CallSite(CI),
+                                 IFI,
+                                 nullptr,
+                                 true);
+
+    // Still need to replace the successor with the fall-through.
+    ReplaceInstWithInst(Term, Br);
+    break;
+  }
+
+  default:
+    revng_abort();
+  }
+
+  Call->eraseFromParent();
+}
+
+template <class FunctionOracle>
+llvm::Function *CFEPAnalyzer<FunctionOracle>::createDisposableFunction(llvm::BasicBlock *Entry) {
+  using namespace llvm;
+  Function *Root = Entry->getParent();
+
+  OnceQueue<BasicBlock *> Queue;
+  std::vector<BasicBlock *> BlocksToCloneVec;
+  Queue.insert(Entry);
+
+  while (!Queue.empty()) {
+    BasicBlock *Current = Queue.pop();
+
+    BlocksToCloneVec.emplace_back(Current);
+
+    if (GCBI::isFunctionCall(Current)) {
+      auto *Succ = GCBI::getFallthrough(Current);
+      revng_assert(Succ != nullptr);
+      Queue.insert(Succ);
+      continue;
+    }
+
+    for (auto *Succ : successors(Current)) {
+      if (!GCBI::isPartOfRootDispatcher(Succ))
+        Queue.insert(Succ);
+    }
+  }
+
+  ValueToValueMapTy VMap;
+  SmallVector<BasicBlock *, 8> BlocksToExtractVec;
+
+  for (const auto &BB : BlocksToCloneVec) {
+    auto *Cloned = CloneBasicBlock(BB, VMap, Twine("__cloned"), Root);
+
+    if (GCBI::isFunctionCall(Cloned))
+      integrateFunctionCallee(Cloned);
+
+    VMap[BB] = Cloned;
+    BlocksToExtractVec.emplace_back(Cloned);
+  }
+
+  auto *First = BlocksToExtractVec.front();
+  auto *NewEntryDummy = BasicBlock::Create(
+                          Root->getContext(),
+                          First->getName() + Twine("__dummyentry"),
+                          Root,
+                          First);
+
+  BranchInst::Create(First, NewEntryDummy);
+  BlocksToExtractVec.insert(BlocksToExtractVec.begin(), NewEntryDummy);
```

Comment:
It turns out that with the dismissal of allocas, a dummy entry basic block is not required any longer. I just removed it all (the shift and insert was indeed inefficient).

Reply:
Wait, you still need it. The entry block cannot have predecessors in LLVM IR. If you have a backedge targeting the entry point you'll have invalid IR.
We have to handle the same situation in function isolation.

Reply:
`CodeExtractor` already provides a dummy entry, which just jumps to the original entry basic block. I used to create a further dummy entry basic block merely for the allocas, since the one provided by `CodeExtractor` doesn't exist yet (the extraction still needs to be done). If the are backward edges to the original entry basic block, it won't be a problem, since a dummy entry basic block is about to be added when extracting the whole function (and this latter will never have any predecessors).

Reply:
Oh OK, didn't know that.